src/ray/gcs/client.cc

```diff
@@ -2,51 +2,150 @@
+    // Sleep for a little, and try again if not all Redis shard addresses have
+    // been added yet.
+    freeReplyObject(reply);
+    usleep(RayConfig::instance().redis_db_connect_wait_milliseconds() * 1000);
+    num_attempts++;
+  }
+  RAY_CHECK(num_attempts < RayConfig::instance().redis_db_connect_retries())
+      << "Expected " << num_redis_shards << " Redis shard addresses, found "
+      << reply->elements;
+
+  // Parse the Redis shard addresses.
+  for (size_t i = 0; i < reply->elements; ++i) {
+    // Parse the shard addresses and ports.
+    RAY_CHECK(reply->element[i]->type == REDIS_REPLY_STRING);
+    std::string addr;
+    std::stringstream ss(reply->element[i]->str);
+    getline(ss, addr, ':');
+    addresses.push_back(addr);
+    int port;
+    ss >> port;
+    ports.push_back(port);
+  }
+  freeReplyObject(reply);
+}
+
 namespace ray {
 
 namespace gcs {
 
-AsyncGcsClient::AsyncGcsClient(const ClientID &client_id, CommandType command_type) {
-  context_ = std::make_shared<RedisContext>();
+AsyncGcsClient::AsyncGcsClient(const std::string &address, int port,
+  const ClientID &client_id, CommandType command_type,
+  bool is_test_client = false) {
   primary_context_ = std::make_shared<RedisContext>();
-  client_table_.reset(new ClientTable(primary_context_, this, client_id));
-  object_table_.reset(new ObjectTable(context_, this, command_type));
-  actor_table_.reset(new ActorTable(context_, this));
-  task_table_.reset(new TaskTable(context_, this, command_type));
-  raylet_task_table_.reset(new raylet::TaskTable(context_, this, command_type));
-  task_reconstruction_log_.reset(new TaskReconstructionLog(context_, this));
-  task_lease_table_.reset(new TaskLeaseTable(context_, this));
-  heartbeat_table_.reset(new HeartbeatTable(context_, this));
-  driver_table_.reset(new DriverTable(primary_context_, this));
-  error_table_.reset(new ErrorTable(primary_context_, this));
-  profile_table_.reset(new ProfileTable(context_, this));
+  // A boolean placeholder value.
+  // Used since changing the API of Connect is breaking some tests.
+  // Would be removed soon in the future.
+  const bool DUMMY_BOOL = true;
+  RAY_CHECK_OK(primary_context_->Connect(address, port, DUMMY_BOOL));
+
+  if (!is_test_client) {
+    // Moving sharding into constructor defaultly means that sharding = true.
+    // This design decision may worth a look.
+    std::vector<std::string> addresses;
+    std::vector<int> ports;
+    GetRedisShards(primary_context_->sync_context(), addresses, ports);
+    if (addresses.size() == 0 || ports.size() == 0) {
+      addresses.push_back(address);
+      ports.push_back(port);
+    }
+
+    // Populate shard_contexts.
+    for (unsigned int i = 0; i < addresses.size(); ++i) {
+      // Slower than emplace but resource safe.
+      shard_contexts_.push_back(std::make_shared<RedisContext>());
+    }
+
+    // Call connect for all contexts. Safe to do many times.
+    // Here shard_contexts_.size() == addresses.size();
```

Comment:
Consider making this comment a `RAY_CHECK(shard_contexts_.size() == addresses.size())`.

Reply:
Yep.